src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -6257,59 +6206,84 @@ private enum ParseTypeMode
             var type = this.ParseUnderlyingType(parentIsParameter: mode == ParseTypeMode.Parameter, options: nameOptions);
             Debug.Assert(type != null);
 
-            if (this.CurrentToken.Kind == SyntaxKind.QuestionToken &&
-                // we do not permit nullable types in a declaration pattern
-                (mode != ParseTypeMode.AfterIs && mode != ParseTypeMode.DefinitePattern || !IsTrueIdentifier(this.PeekToken(1))))
-            {
-                var question = EatNullableQualifierIfApplicable(mode);
-                if (question != null)
-                {
-                    type = _syntaxFactory.NullableType(type, question);
-                }
-            }
-
-            switch (mode)
-            {
-                case ParseTypeMode.AfterIs:
-                case ParseTypeMode.DefinitePattern:
-                case ParseTypeMode.AfterTupleComma:
-                case ParseTypeMode.FirstElementOfPossibleTupleLiteral:
-                    // these contexts do not permit a pointer type except as an element type of an array.
-                    if (PointerTypeModsFollowedByRankAndDimensionSpecifier())
-                    {
-                        type = this.ParsePointerTypeMods(type);
-                    }
-                    break;
-                case ParseTypeMode.Normal:
-                case ParseTypeMode.Parameter:
-                case ParseTypeMode.AfterOut:
-                case ParseTypeMode.ArrayCreation:
-                case ParseTypeMode.AsExpression:
-                    type = this.ParsePointerTypeMods(type);
-                    break;
-            }
-
-            // Now check for arrays.
-            if (this.IsPossibleRankAndDimensionSpecifier())
+            int lastTokenPosition = -1;
+            while (IsMakingProgress(ref lastTokenPosition))
             {
-                var ranks = _pool.Allocate<ArrayRankSpecifierSyntax>();
-                try
+                switch (this.CurrentToken.Kind)
                 {
-                    while (this.IsPossibleRankAndDimensionSpecifier())
-                    {
-                        bool unused;
-                        var rank = this.ParseArrayRankSpecifier(mode == ParseTypeMode.ArrayCreation, expectSizes, questionTokenModeOpt: mode, out unused);
-                        ranks.Add(rank);
-                        expectSizes = false;
-                    }
+                    case SyntaxKind.QuestionToken when canBeNullableType():
+                        {
+                            var question = EatNullableQualifierIfApplicable(mode);
+                            if (question != null)
+                            {
+                                type = _syntaxFactory.NullableType(type, question);
+                                continue;
+                            }
+                            goto done; // token not consumed
+                        }
+                        bool canBeNullableType()
```

Comment:
>bool canBeNullableType() [](start = 24, length = 24)

nit: we've favored the convention of putting local functions at the end of methods

Reply:
I don't think it's horrible, as it is indeed in a dead space, so does not interfere much with readability of code flow. 
But when I read the `when` clause I scrolled down to the end of the method to look for it, then scrolled back up when I could not find it. This may just be an artifact of limited diff tool, which won't be a problem in the IDE (GoToDef).
That's why I think the end of the method is a valuable convention, and is better than a "dead code" convention.

---
In reply to: [248809704](https://github.com/dotnet/roslyn/pull/32431#discussion_r248809704) [](ancestors = 248809704,248783958)